File: 08-composite/08-composite.php
<?php

interface Item
{
    public function package(): void;
    public function add(Item $item): void;
}

class Product implements Item
{
    public function __construct(private string $name)
    {
    }
    public function package(): void
    {
        echo "- " . $this->name . "\n";
    }

    public function add(Item $item): void
    {
        throw new Exception("Cannot add items to a product.");
    }
}

class Box implements Item
{
    private array $children = [];
    public function __construct(private string $name)
    {
    }
    public function add(Item $item): void
    {
        $this->children[] = $item;
    }
    public function package(): void
    {
        echo "+ " . $this->name . "\n";

        foreach ($this->children as $child) {
            $child->package();
        }
    }
}

$p1 = new Product("Book");
$p2 = new Product("Pen");

$box1 = new Box("Gift Wrap");

$box1->add($p1);
$box1->add($p2);

$box1->package();

File: 08-composite/08-composite.py
from abc import ABC, abstractmethod

class Item(ABC):
    @abstractmethod
    def package(self):
        pass

class Product(Item):
    def __init__(self, title):
        self.title = title

    def package(self):
        print(f"  - {self.title}")

class Box(Item):
    def __init__(self, title):
        self.title = title
        self.children = []

    def package(self):
        print(f"Box: {self.title}")
        for child in self.children:
            child.package()

    def add(self, item):
        self.children.append(item)


p1 = Product("car")
p2 = Product("bike")

b = Box("garage")
b.add(p1)
b.add(p2)

b.package()
